src/domain/checkForAndPlaceOrder.ts

```diff
@@ -659,19 +665,24 @@ function _handleGetTradableOrderWithSignatureCall(
           };
       }
     } catch (err: any) {
-      log.error(`${logPrefix} Unexpected error`, err);
+      // Any errors thrown here can _ONLY_ come from non-compliant interfaces (ie. bad revert ABI encoding).
+      // We log the error, and return a DONT_TRY_AGAIN result.
+      // TODO: Add metrics to track this
+      log.error(
+        `Contract returned non-interface compliant revert via getTradeableOrderWithSignature. Simulate: https://dashboard.tenderly.co/gp-v2/watch-tower-prod/simulator/new?network=${chainId}&contractAddress=${to}&rawFunctionInput=${data}`
+      );
       return {
-        result: PollResultCode.UNEXPECTED_ERROR,
+        result: PollResultCode.DONT_TRY_AGAIN,
```

Comment:
I don't think this is a good idea. catching ANY error thrown in the block above and just stop indexing the order might be a source of bugs.

Can't we handle it in a different way? 

Maybe could be usefult to extract the error handling to a testable function and create some tests. This way, we can make sure we capture the cases you want to capture and there's no future regressions. 

Also, by making the error handling independent, it will make `checkForAndPlaceOrders` thinner, which would be a good thing


Alternativelly, we could move the whole legacy polling and do the same. This in fact is something that is good is independent, since it should be moved to the SDK. Watch Tower logic should be simpler

Reply:
> The only code path here (yes, it should be tested) that should throw and result in invalidating the order, is customErrorDecode that solely depends on only ABI decoding

I would hope we structure it so typescript already can check this assumption 

> Comprehensive logging. We have ELK in place, and much work has been done with logging to ensure great visibility.
Comprehensive metrics. In https://github.com/cowprotocol/watch-tower/pull/95 the order tuple (chainId, handler, owner, id) is labelled against the watch_tower_polling_onchain_invalid_interface_total metric. Therefore we can see when orders are getting dropped due to this, and investigate to ensure compliance.

I think these don't solve the problem. I think it should be impossible to remove an order if there's some other error that is not the decoding. Typescript can probably help with these at compile time

> We do some unit testing - https://github.com/cowprotocol/watch-tower/issues/97 - and validate while in prod.

This will help indeed

Reply:
I've gone through and implemented everything with very strong typing now to enforce the assumptions. Some tests have been added as well.

Reply:
Looks better, but you are still not exhaustive. You can't be. Because you use `parsedCustomError.selector` in the switch, you can't know all selectors. You need to handle an unknown selector, but your code now assumes its impossible that will happen

Reply:
The revert data is parsed in `parseCustomError` to ensure that it complies with the ComposableCoW interface. If it does not, it will throw. Therefore the switch is exhaustive as only valid selectors reach that point in the execution path.

Invalid selectors (ie. that do _not_ comply with the interface that we have prescribed), are caught, dropped, and metrics are in place to monitor this.